### PR TITLE
Replace LGPL inflate code with miniz

### PR DESCRIPTION
**Pull request by ccawley2011**
This is not yet finished; the gzip depacker has not been adapted yet and some of the MUSE tests are failing.

### CODE REVIEW

**Comment by AliceLR** [comment]
It looks like the three failing tests are all fuzzer files, so they probably have invalid DEFLATE streams that the old inflate implementation was fine with—I can submit a separate patch to extract them since the container isn't relevant to what is being tested.

**Comment by sezero** [comment]
[rant] Gotta love single-header stuff.. All in one place: OK, good.
Too-much-in-one-place: ugh.. Takes-no-less-than-two-days-to-
navigate-the-header-because-they-are-10km-long: ugh ugh eww. [/rant]

That aside: here are a few things after a very brief look:

- miniz tries very hard to use 64 bit file offsets, which we
  don't need as far as I know.

- it uses oh-so-very-secure fopen_s and freopen_s in case of
  msvc (and mingw-w64 with its incorrect detection): we don't
  need. not all versions may have them either.

- makes assumptions about toolchains, e.g.:

- it defaults to fseeko and ftello unless it's in ansi mode:
  fseeko/ftello needn't be available at all, e.g. with watcom
  or with some other toolchains/targets I have. So: We either
  should plainly use fseek/ftell, or at least detect fseeko &
  ftello at configure time.

- it packs a lot of functionality many of which we don't need.
  some of them are configurable: we should use.  some of them
  are not, e.g. compressing facilities as far as I can see, we
  should customize to keep them out.

Those aside: You should rebase to master, should fix the muse
test failures.

**Comment by sezero** [comment]
Also, when this settles, we should 'undepackerize' muse.c and move to loaders.

**Comment by sezero** [comment]
Also got one warning:
```
src/depackers/miniz.c: In function ‘mz_bool mz_zip_validate_file(mz_zip_archive*, mz_uint, mz_uint)’:
src/depackers/miniz.c:5353: warning: dereferencing type-punned pointer will break strict-aliasing rules
```

**Comment by sezero** [comment]
And here are watcom W007 warnings in c++ mode (don't know how to detect
those with gcc.)
```
src/depackers/miniz.c(747): Warning! W007: col(22) '&array' may not produce intended result
src/depackers/miniz.c(865): Warning! W007: col(27) '&array' may not produce intended result
src/depackers/miniz.c(891): Warning! W007: col(53) '&array' may not produce intended result
src/depackers/miniz.c(892): Warning! W007: col(48) '&array' may not produce intended result
src/depackers/miniz.c(1945): Warning! W007: col(35) '&array' may not produce intended result
src/depackers/miniz.c(1946): Warning! W007: col(35) '&array' may not produce intended result
src/depackers/miniz.c(1969): Warning! W007: col(31) '&array' may not produce intended result
src/depackers/miniz.c(1990): Warning! W007: col(31) '&array' may not produce intended result
src/depackers/miniz.c(2512): Warning! W007: col(56) '&array' may not produce intended result
src/depackers/miniz.c(2527): Warning! W007: col(40) '&array' may not produce intended result
src/depackers/miniz.c(2528): Warning! W007: col(47) '&array' may not produce intended result
src/depackers/miniz.c(2529): Warning! W007: col(44) '&array' may not produce intended result
src/depackers/miniz.c(5791): Warning! W007: col(25) '&array' may not produce intended result
src/depackers/miniz.c(6247): Warning! W007: col(34) '&array' may not produce intended result
src/depackers/miniz.c(6500): Warning! W007: col(34) '&array' may not produce intended result
src/depackers/miniz.c(7287): Warning! W007: col(25) '&array' may not produce intended result
src/depackers/miniz.c(7302): Warning! W007: col(25) '&array' may not produce intended result
src/depackers/miniz.c(7313): Warning! W007: col(21) '&array' may not produce intended result
```

**Comment by sezero** [comment]
An easy way to silence those may be defining a `MZ_CLEAR_ARR` without the `&`
and change the offending `MZ_CLEAR_OBJ` with it

**Comment by sezero** [comment]
Here is a tiny patch that fixes those W007 warnings: [patch-W007.txt](https://github.com/libxmp/libxmp/files/6848868/patch-W007.txt)

**Comment by sezero** [comment]
FWIW, this PR https://github.com/richgel999/miniz/pull/159 in mainstream looks promising? (Haven't tested myself yet.)

File: src/depackers/miniz.h [hunk 1]
```diff
@@ -0,0 +1,1356 @@
...
+
+#if defined(_M_IX86) || defined(_M_X64) || defined(__i386__) || defined(__i386) || defined(__i486__) || defined(__i486) || defined(i386) || defined(__ia64__) || defined(__x86_64__)
+/* MINIZ_X86_OR_X64_CPU is only used to help set the below macros. */
+#define MINIZ_X86_OR_X64_CPU 1
+#else
+#define MINIZ_X86_OR_X64_CPU 0
+#endif
+
+#if (__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__) || MINIZ_X86_OR_X64_CPU
+/* Set MINIZ_LITTLE_ENDIAN to 1 if the processor is little endian. */
```
Comment (sezero) [inline src/depackers/miniz.h:166]: This naively assumes gcc >= 6.x - or its x86 definition: very fragile and
has an in-negligible possibility of being wrong.  We should use our own
configury define `WORDS_BIGENDIAN` here.

Also see mainstream issue: https://github.com/richgel999/miniz/issues/166

Comment (sezero) [reply src/depackers/miniz.h:166]: > This naively assumes gcc >= 6.x 

correction: that would be gcc >= 4.6.x

**Comment by sezero** [comment]
- Need rebasing.
- Do we want  fopen_s and freopen_s things in windows builds?
- Do want 64 bit offsets?
- The watcom W007 patch isn't in.
- I still get `src/depackers/miniz.c:5369: warning: dereferencing type-punned pointer will break strict-aliasing rules`
  (not necessarily a serious one.)

Ran test/test.c under valgrind:
`==31726==   total heap usage: 42 allocs, 42 frees, 251,627 bytes allocated` versus
` ==1709==   total heap usage: 44 allocs, 44 frees, 273,111 bytes allocated`
i.e. slightly higher memory usage.

Is there an easy way to check stack usage?

**Comment by sezero** [comment]
Also we have this failure:
https://github.com/libxmp/libxmp/pull/436/checks?check_run_id=3125395583#step:7:1431
```
Uninitialized bytes in __interceptor_fwrite at offset 0 inside [0x72c000000000, 1314)
==6226==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f89adb983cf in tinfl_put_buf_func /home/runner/work/libxmp/libxmp/src/depackers/gunzip.c:53:21
    #1 0x7f89adb04dfd in tinfl_decompress_mem_to_callback /home/runner/work/libxmp/libxmp/src/depackers/miniz.c:2940:33
    #2 0x7f89adb97b77 in decrunch_gzip /home/runner/work/libxmp/libxmp/src/depackers/gunzip.c:138:6
    #3 0x7f89adaa6b45 in libxmp_decrunch /home/runner/work/libxmp/libxmp/src/depackers/depacker.c:302:7
    #4 0x7f89ad69d3df in xmp_load_module /home/runner/work/libxmp/libxmp/src/load.c:372:6
    #5 0x68cc4a in _test_func_test_fuzzer_depack_zip_truncated2 /home/runner/work/libxmp/libxmp/test-dev/test_fuzzer_depack_zip_truncated2.c:13:8
    #6 0x4b247a in run_tests /home/runner/work/libxmp/libxmp/test-dev/main.c:95:9
    #7 0x4b9c71 in main /home/runner/work/libxmp/libxmp/test-dev/main.c:173:6
    #8 0x7f89ad3150b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #9 0x41d5dd in _start (/home/runner/work/libxmp/libxmp/test-dev/libxmp-tests+0x41d5dd)

  Uninitialized value was stored to memory at
    #0 0x7f89adafc295 in tinfl_decompress /home/runner/work/libxmp/libxmp/src/depackers/miniz.c:2780:37
    #1 0x7f89adb04b54 in tinfl_decompress_mem_to_callback /home/runner/work/libxmp/libxmp/src/depackers/miniz.c:2937:31
    #2 0x7f89adb97b77 in decrunch_gzip /home/runner/work/libxmp/libxmp/src/depackers/gunzip.c:138:6
    #3 0x7f89adaa6b45 in libxmp_decrunch /home/runner/work/libxmp/libxmp/src/depackers/depacker.c:302:7
    #4 0x7f89ad69d3df in xmp_load_module /home/runner/work/libxmp/libxmp/src/load.c:372:6
    #5 0x68cc4a in _test_func_test_fuzzer_depack_zip_truncated2 /home/runner/work/libxmp/libxmp/test-dev/test_fuzzer_depack_zip_truncated2.c:13:8
    #6 0x4b247a in run_tests /home/runner/work/libxmp/libxmp/test-dev/main.c:95:9
    #7 0x4b9c71 in main /home/runner/work/libxmp/libxmp/test-dev/main.c:173:6
    #8 0x7f89ad3150b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

  Uninitialized value was created by a heap allocation
    #0 0x429a1d in malloc (/home/runner/work/libxmp/libxmp/test-dev/libxmp-tests+0x429a1d)
    #1 0x7f89adb0448b in tinfl_decompress_mem_to_callback /home/runner/work/libxmp/libxmp/src/depackers/miniz.c:2929:35
    #2 0x7f89adb97b77 in decrunch_gzip /home/runner/work/libxmp/libxmp/src/depackers/gunzip.c:138:6
    #3 0x7f89adaa6b45 in libxmp_decrunch /home/runner/work/libxmp/libxmp/src/depackers/depacker.c:302:7
    #4 0x7f89ad69d3df in xmp_load_module /home/runner/work/libxmp/libxmp/src/load.c:372:6
    #5 0x68cc4a in _test_func_test_fuzzer_depack_zip_truncated2 /home/runner/work/libxmp/libxmp/test-dev/test_fuzzer_depack_zip_truncated2.c:13:8
    #6 0x4b247a in run_tests /home/runner/work/libxmp/libxmp/test-dev/main.c:95:9
    #7 0x4b9c71 in main /home/runner/work/libxmp/libxmp/test-dev/main.c:173:6
    #8 0x7f89ad3150b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
```

**Comment by sezero** [comment]
> Ran test/test.c under valgrind:
> `==31726== total heap usage: 42 allocs, 42 frees, 251,627 bytes allocated` versus
> ` ==1709== total heap usage: 44 allocs, 44 frees, 273,111 bytes allocated`
> i.e. slightly higher memory usage.

Tried applying https://github.com/richgel999/miniz/pull/159, didn't change anything for this particular testcase.

**Comment by sezero** [comment]
OK, here is the stack usage analysis from gcc4.9.4 `-fstack-usage`.  Old:
```
gunzip.c:44:12:test_gzip	4	static
gunzip.c:49:12:decrunch_gzip	64	dynamic,bounded

muse.c:26:12:decrunch_muse	48	dynamic,bounded
muse.c:13:12:test_muse	32	dynamic,bounded

unzip.c:283:12:test_zip	4	static
unzip.c:126:12:read_zip_header	32	dynamic,bounded
unzip.c:44:12:read_chars	48	dynamic,bounded
unzip.c:226:12:kunzip_get_offset_excluding	1152	dynamic,bounded
unzip.c:97:12:copy_file.isra.0	16816	dynamic,bounded
unzip.c:290:12:decrunch_zip	112	dynamic,bounded

inflate.c:149:12:reverse_bitstream	24	static
inflate.c:169:12:add_static_codes_to_tree	36	static
inflate.c:287:12:load_codes	4160	dynamic,bounded
inflate.c:712:5:decompress	64	dynamic,bounded
inflate.c:1021:5:libxmp_inflate	34496	dynamic,bounded
```

Versus new:
```
gunzip.c:46:12:test_gzip	4	static
gunzip.c:56:12:decrunch_gzip	80	dynamic,bounded
gunzip.c:51:12:tinfl_put_buf_func	32	dynamic,bounded

muse.c:45:12:decrunch_muse	80	dynamic,bounded
muse.c:40:12:tinfl_put_buf_func	32	dynamic,bounded
muse.c:27:12:test_muse	32	dynamic,bounded

unzip.c:27:12:test_zip	4	static
unzip.c:52:12:decrunch_zip	672	dynamic,bounded
unzip.c:43:15:mz_zip_file_write_callback	32	dynamic,bounded
unzip.c:35:15:mz_zip_file_read_func	32	dynamic,bounded

miniz.c:172:19:miniz_def_free_func	32	dynamic,bounded
miniz.c:167:20:miniz_def_alloc_func	32	dynamic,bounded
miniz.c:177:20:miniz_def_realloc_func	32	dynamic,bounded
miniz.c:3895:15:mz_zip_mem_read_func	80	dynamic,bounded
miniz.c:3824:16:mz_zip_reader_end_internal	32	dynamic,bounded
miniz.c:3405:16:mz_zip_reader_init_internal	48	dynamic,bounded
miniz.c:3276:16:mz_zip_array_ensure_capacity.isra.5.constprop	48	dynamic,bounded
miniz.c:3566:16:mz_zip_reader_read_central_dir	4336	dynamic,bounded
miniz.c:40:10:mz_adler32	48	static
miniz.c:95:10:mz_crc32	24	static
miniz.c:5199:15:mz_zip_compute_crc32_callback	24	dynamic,bounded
miniz.c:162:6:mz_free	32	dynamic,bounded
miniz.c:183:13:mz_version	4	static
miniz.c:2405:14:tinfl_decompress	272	dynamic,bounded
miniz.c:2878:7:tinfl_decompress_mem_to_heap	11088	dynamic,bounded
miniz.c:2916:8:tinfl_decompress_mem_to_mem	11040	dynamic,bounded
miniz.c:2925:5:tinfl_decompress_mem_to_callback	11104	dynamic,bounded
miniz.c:2955:21:tinfl_decompressor_alloc	32	dynamic,bounded
miniz.c:2963:6:tinfl_decompressor_free	32	dynamic,bounded
miniz.c:3818:6:mz_zip_zero_struct	8	static
miniz.c:3871:9:mz_zip_reader_end	4	static
miniz.c:3875:9:mz_zip_reader_init	16	static
miniz.c:3903:9:mz_zip_reader_init_mem	16	static
miniz.c:4059:9:mz_zip_reader_is_file_encrypted	8	static
miniz.c:4073:9:mz_zip_reader_is_file_supported	8	static
miniz.c:4109:9:mz_zip_reader_is_file_a_directory	8	static
miniz.c:4141:16:mz_zip_file_stat_internal.part.8	80	dynamic,bounded
miniz.c:4338:9:mz_zip_reader_locate_file_v2	112	dynamic,bounded
miniz.c:4329:5:mz_zip_reader_locate_file	64	dynamic,bounded
miniz.c:4967:8:mz_zip_reader_extract_iter_read	112	dynamic,bounded
miniz.c:5084:9:mz_zip_reader_extract_iter_free	32	dynamic,bounded
miniz.c:7529:13:mz_zip_get_mode	4	static
miniz.c:7534:13:mz_zip_get_type	4	static
miniz.c:7539:14:mz_zip_set_last_error	4	static
miniz.c:7552:14:mz_zip_peek_last_error	4	static
miniz.c:7560:14:mz_zip_clear_last_error	4	static
miniz.c:7565:14:mz_zip_get_last_error	4	static
miniz.c:7578:13:mz_zip_get_error_string	4	static
miniz.c:7654:9:mz_zip_is_zip64	4	static
miniz.c:7662:8:mz_zip_get_central_dir_size	4	static
miniz.c:7670:9:mz_zip_reader_get_num_files	4	static
miniz.c:7675:11:mz_zip_get_archive_size	4	static
miniz.c:7682:11:mz_zip_get_archive_file_start_offset	4	static
miniz.c:7689:10:mz_zip_get_cfile	4	static
miniz.c:7696:8:mz_zip_read_archive_data	48	dynamic,bounded
miniz.c:7704:9:mz_zip_reader_get_filename	32	dynamic,bounded
miniz.c:7725:9:mz_zip_reader_file_stat	12	static
miniz.c:4403:9:mz_zip_reader_extract_to_mem_no_alloc.part.9	12240	dynamic,bounded
miniz.c:4403:9:mz_zip_reader_extract_to_mem_no_alloc	32	static
miniz.c:4546:9:mz_zip_reader_extract_file_to_mem_no_alloc	64	dynamic,bounded
miniz.c:4559:9:mz_zip_reader_extract_file_to_mem	48	dynamic,bounded
miniz.c:4554:9:mz_zip_reader_extract_to_mem	48	dynamic,bounded
miniz.c:4564:7:mz_zip_reader_extract_to_heap	64	dynamic,bounded
miniz.c:4606:7:mz_zip_reader_extract_file_to_heap	64	dynamic,bounded
miniz.c:4618:9:mz_zip_reader_extract_to_callback.part.10	12256	dynamic,bounded
miniz.c:4618:9:mz_zip_reader_extract_to_callback	12	static
miniz.c:4818:9:mz_zip_reader_extract_file_to_callback	64	dynamic,bounded
miniz.c:5207:9:mz_zip_validate_file	1312	dynamic,bounded
miniz.c:4827:35:mz_zip_reader_extract_iter_new	112	dynamic,bounded
miniz.c:4955:35:mz_zip_reader_extract_file_iter_new	64	dynamic,bounded
miniz.c:5422:9:mz_zip_validate_archive	1168	dynamic,bounded
miniz.c:5475:9:mz_zip_validate_mem_archive	112	dynamic,bounded
miniz.c:7730:9:mz_zip_end	4	static
```

**Comment by sezero** [comment]
https://github.com/d99kris/stackusage reports 46044 to 53180 for original
```
stackusage log at 2021-07-22 14:07:28 ----------------------------------------
  pid  id    tid  requested     actual     maxuse  max%    dur               funcP name
17940   0  17940   10485760   10485760      53180     0      0               (nil) libxmp-test
```
versus 25708 to 28140 with miniz, which is definitely an improvement:
```
stackusage log at 2021-07-22 14:08:15 ----------------------------------------
  pid  id    tid  requested     actual     maxuse  max%    dur               funcP name
18036   0  18036   10485760   10485760      28140     0      0               (nil) libxmp-test
```

**Comment by sezero** [comment]
Uninitialized memory issue is still there:
https://github.com/libxmp/libxmp/pull/436/checks?check_run_id=3134336544#step:7:1433

Or, from my local run under valgrind:
```
Loading depack_zip_truncated2.zip (1 of 1)
==23206== Syscall param write(buf) points to uninitialised byte(s)
==23206==    at 0x9A0CD3: __write_nocancel (syscall-template.S:82)
==23206==    by 0x93BA8F: _IO_do_write@@GLIBC_2.1 (fileops.c:522)
==23206==    by 0x93B20F: _IO_file_close_it@@GLIBC_2.1 (fileops.c:169)
==23206==    by 0x92E6D7: fclose@@GLIBC_2.1 (iofclose.c:62)
==23206==    by 0x4067400: libxmp_decrunch (depacker.c:321)
==23206==    by 0x4024B37: xmp_load_module (load.c:372)
==23206==    by 0x804B120: main (main.c:399)
==23206==  Address 0x4091000 is not stack'd, malloc'd or (recently) free'd
```

Lazy work-around is zeroing the malloc()ed buffer like
```diff
diff --git a/src/depackers/miniz.c b/src/depackers/miniz.c
index 924601c..e492f55 100644
--- a/src/depackers/miniz.c
+++ b/src/depackers/miniz.c
@@ -2931,4 +2931,5 @@ int tinfl_decompress_mem_to_callback(const void *pIn_buf, size_t *pIn_buf_size,
     if (!pDict)
         return TINFL_STATUS_FAILED;
+    memset(pDict,0,TINFL_LZ_DICT_SIZE);
     tinfl_init(&decomp);
     for (;;)
```
Anything better?

**Comment by sezero** [comment]
#AliceLR: Please have a look at this.

It looks pretty well, except for the uninitialized memory issue --
Workaround: https://github.com/libxmp/libxmp/pull/436#issuecomment-884977962

I (for myself), don't like the 64 bit offsets use much, but won't
fight for it.

**Comment by AliceLR** [comment]
A large portion of this library (thousands of lines in several continuous blocks) seems to be related to archive creation/PNG decoding/???, i.e. completely useless to libxmp. This is all deadcode and, unlike the stb_vorbis encoding functions, is the vast majority of this library and very self contained. IMO just remove all of this.

**Comment by AliceLR** [comment]
> It looks pretty well, except for the uninitialized memory issue --
> Workaround: [#436 (comment)](https://github.com/libxmp/libxmp/pull/436#issuecomment-884977962)

I'm not 100% sure what's going on here (this is way too much to sift through in its current state) but my initial impression is it's referencing junk data in the sliding dictionary. This is strictly *wrong* behavior—the sliding dictionary should be treated as zero initialized, so your patch is correct (better would be if they bothered to provide an `MZ_CALLOC`, but whatever).

**Review by AliceLR — CHANGES_REQUESTED** [review]
I think pulling in the entirety of miniz's source code, including compression, archive writing, and PNG(??) routines is unnecessary when tinfl.c is available as a standalone file. Using just tinfl.c would leave unzip.c as LGPL, but unzip.c does nothing that isn't fairly easy to reimplement later (without pulling 8000 lines of mostly useless code into libxmp's source).

**Comment by sezero** [comment]
> I think pulling in the entirety of miniz's source code, including compression, archive writing, and PNG(??)

Heh-he - see my rant at the top!

**Comment by ccawley2011** [comment]
CI now passes, and all of the changes have been submitted upstream except for the libxmp-specific section near the top of `miniz.h`.

> I think pulling in the entirety of miniz's source code, including compression, archive writing, and PNG(??) routines is unnecessary when tinfl.c is available as a standalone file. Using just tinfl.c would leave unzip.c as LGPL, but unzip.c does nothing that isn't fairly easy to reimplement later (without pulling 8000 lines of mostly useless code into libxmp's source).

The [readme file](https://github.com/richgel999/miniz#usage) specifically discourages using the non-amalgamated files, so I'm not sure this is a good idea. This PR already disables the unused deflate, archive writing and stdio code, so this shouldn't cause much of an impact on the resulting binary.

**Comment by AliceLR** [comment]
> The [readme file](https://github.com/richgel999/miniz#usage) specifically discourages using the non-amalgamated files

What I mean is miniz [explicitly distributes a standalone tinfl implementation](https://github.com/richgel999/miniz/blob/master/miniz_tinfl.c). From their readme:

> Entire inflater (including optional zlib header parsing and Adler-32 checking) is implemented in a single function as a coroutine, which is separately available in a small (~550 line) source file: miniz_tinfl.c

**Comment by sezero** [comment]
The attached patch aggresively strips down miniz - don't know whether
anyone would agree or disagree with it, but here it is.  With this,
miniz.c and miniz.h goes down to 100kb and 31kb, and the compiled *.so
is about 8.5kb smaller (depends on the compiler, etc, I guess.)  Here
is what it does:

* remove everything that libxmp-specific config already ifdefs-out.

* remove the following procs not used in libxmp:
  all to_heap, to_mem, to_file and to_cfile procs along with their
  dependencies, tinfl_decompressor_alloc|free,  iterative extraction
  procs, mz_zip_get_archive_file_start_offset, mz_zip_validate procs
  and their dependencies, mz_zip_zero_struct, mz_zip_array_init,
  mz_zip_array_ensure_room, mz_zip_array_reserve, mz_zip_array_push_back,
  mz_adler32, mz_version.

* comment out some of the procs in 'Misc utils' section.

* remove m_pFile, m_file_archive_start_ofs and also m_mem_size and
  m_mem_capacity from struct mz_zip_internal_state_tag.

* make miniz_def_alloc_func & co, and mz_crc32 static to miniz.c.

[patch_miniz_strip.txt](https://github.com/libxmp/libxmp/files/6885876/patch_miniz_strip.txt)

**Comment by ccawley2011** [comment]
The issue I have is that I don't want to introduce too many invasive changes to miniz, as I want to avoid the situation with PR #408 where updating stb_vorbis is complicated by the number of changes that need to be applied to the new version.

**Comment by sezero** [comment]
@AliceLR,  @cmatsuoka ?

**Comment by sezero** [comment]
Side note: we may define NDEBUG in our libxmp-specific configuration, I guess.

**Comment by AliceLR** [comment]
To clarify, I'm not advocating for stripping down miniz and I don't think that's a good idea. I'm advocating for dropping the full version of miniz entirely and using tinfl.c by itself, which miniz *lists as a supposed feature* (despite also claiming that you shouldn't use non-amalgamated files?? The original Google Code site for miniz contains no such contradictory claim though...).

**Comment by sezero** [comment]
Mandatory rebase reminder: `cmake/libxmp-sources.cmake` should be modified when rebasing.

**Comment by sezero** [comment]
What are we going to do about this?

**Comment by sezero** [comment]
PING?

**Comment by sezero** [comment]
Rebase? PING?

**Comment by sezero** [comment]
PING^n ?

**Comment by sezero** [comment]
This one needs rebasing after #493 merge.

**Comment by ccawley2011** [comment]
> This one needs rebasing after #493 merge.

OK, done.

**Review by sezero — APPROVED** [review]
OK, let's merge this and _then_ improve upon it later.

**Pull request merged**